File: receivers/common/web/ConnectionMonitor.ts
import { Opcode } from 'common/Packets';
import { Logger, LoggerType } from 'common/Logger';

// Required for webOS since preload declared interface is not available on the backend
declare global {
    interface Window {
      targetAPI: any;
    }
}

// Window might be re-created while devices are still connected
export function setUiUpdateCallbacks(callbacks: any) {
    const logger = window.targetAPI.logger;
    let frontendConnections = [];

    window.targetAPI.onConnect((_event, value: any) => {
        const idMapping = value.type === 'ws' ? value.sessionId : value.data.address;

        logger.debug(`Processing connect event for ${idMapping} with current connections:`, frontendConnections);
        frontendConnections.push(idMapping);
        callbacks.onConnect(frontendConnections);
    });
    window.targetAPI.onDisconnect((_event, value: any) => {
        const idMapping = value.type === 'ws' ? value.sessionId : value.data.address;

        logger.debug(`Processing disconnect event for ${idMapping} with current connections:`, frontendConnections);
        const index = frontendConnections.indexOf(idMapping);
        if (index != -1) {
            frontendConnections.splice(index, 1);
            callbacks.onDisconnect(frontendConnections);
        }
    });

    window.targetAPI.getSessions().then((sessions: string[]) => {
        logger.info('Window created with current sessions:', sessions);
        frontendConnections = sessions;

        if (frontendConnections.length > 0) {
            callbacks.onConnect(frontendConnections, true);
        }
    });
}

export class ConnectionMonitor {
    private static logger: Logger;
    private static connectionPingTimeout = 2500;
    private static heartbeatRetries = new Map<string, number>();
    private static backendConnections = new Map<string, any>(); // is `ListenerService`, but cant import backend module in frontend
    private static uiConnectUpdateTimeout = 100;
    private static uiDisconnectUpdateTimeout = 2000; // Senders may reconnect, but generally need more time
    private static uiUpdateMap = new Map<string, any>(); // { event: string, uiUpdateCallback: () => void }

    constructor() {
        ConnectionMonitor.logger = new Logger('ConnectionMonitor', LoggerType.BACKEND);

        setInterval(() => {
            if (ConnectionMonitor.backendConnections.size > 0) {
                for (const sessionId of ConnectionMonitor.backendConnections.keys()) {
                    const listener = ConnectionMonitor.backendConnections.get(sessionId);

                    if (listener.getSessionProtocolVersion(sessionId) >= 2) {
                        if (ConnectionMonitor.heartbeatRetries.get(sessionId) > 3) {
                            ConnectionMonitor.logger.warn(`Could not ping device with connection id ${sessionId}. Disconnecting...`);
                            listener.disconnect(sessionId);
                        }

                        ConnectionMonitor.logger.debug(`Pinging session ${sessionId} with ${ConnectionMonitor.heartbeatRetries.get(sessionId)} retries left`);
                        listener.send(Opcode.Ping, null, sessionId);
                        ConnectionMonitor.heartbeatRetries.set(sessionId, ConnectionMonitor.heartbeatRetries.get(sessionId) + 1);
                    }
                    else if (listener.getSessionProtocolVersion(sessionId) === undefined) {
                        ConnectionMonitor.logger.warn(`Session ${sessionId} was not found in the list of active sessions. Removing...`);
                        ConnectionMonitor.backendConnections.delete(sessionId);
                        ConnectionMonitor.heartbeatRetries.delete(sessionId);
                    }
                }
            }
        }, ConnectionMonitor.connectionPingTimeout);
    }

    public static onPingPong(sessionId: string, isWebsockets: boolean) {
        ConnectionMonitor.logger.debug(`Received response from ${sessionId}`);

        // Websocket clients currently don't support ping-pong commands
        if (!isWebsockets) {
            ConnectionMonitor.heartbeatRetries.set(sessionId, 0);
        }
    }

    public static onConnect(listener: any, value: any, isWebsockets: boolean, uiUpdateCallback: any) {
        ConnectionMonitor.logger.info(`Device connected: ${JSON.stringify(value)}`);
        const idMapping = isWebsockets ? value.sessionId : value.data.address;

        if (!isWebsockets) {
            ConnectionMonitor.backendConnections.set(value.sessionId, listener);
            ConnectionMonitor.heartbeatRetries.set(value.sessionId, 0);
        }

        // Occasionally senders seem to instantaneously disconnect and reconnect, so suppress those ui updates
        const senderUpdateQueue = ConnectionMonitor.uiUpdateMap.get(idMapping) ?? [];
        senderUpdateQueue.push({ event: 'connect', uiUpdateCallback: uiUpdateCallback });
        ConnectionMonitor.uiUpdateMap.set(idMapping, senderUpdateQueue);

        if (senderUpdateQueue.length === 1) {
            setTimeout(() => { ConnectionMonitor.processUiUpdateCallbacks(idMapping); }, ConnectionMonitor.uiConnectUpdateTimeout);
        }
    }

    public static onDisconnect(value: any, isWebsockets: boolean, uiUpdateCallback: any) {
        ConnectionMonitor.logger.info(`Device disconnected: ${JSON.stringify(value)}`);

        if (!isWebsockets) {
            ConnectionMonitor.backendConnections.delete(value.sessionId);
            ConnectionMonitor.heartbeatRetries.delete(value.sessionId);
        }

        const idMapping = isWebsockets ? value.sessionId : value.data.address;
        const senderUpdateQueue = ConnectionMonitor.uiUpdateMap.get(idMapping);
        senderUpdateQueue.push({ event: 'disconnect', uiUpdateCallback: uiUpdateCallback });
        ConnectionMonitor.uiUpdateMap.set(idMapping, senderUpdateQueue);

        if (senderUpdateQueue.length === 1) {
            setTimeout(() => { ConnectionMonitor.processUiUpdateCallbacks(idMapping); }, ConnectionMonitor.uiDisconnectUpdateTimeout);
        }
    }

    private static processUiUpdateCallbacks(mapId: string) {
        const updateQueue = ConnectionMonitor.uiUpdateMap.get(mapId);
        let lastConnectCb: any;
        let lastDisconnectCb: any;
        let messageCount = 0;

        updateQueue.forEach(update => {
            ConnectionMonitor.logger.debug(`Processing update event '${update.event}' for ${mapId}`);
            if (update.event === 'connect') {
                messageCount += 1;
                lastConnectCb = update.uiUpdateCallback;
            }
            else if (update.event === 'disconnect') {
                messageCount -= 1;
                lastDisconnectCb = update.uiUpdateCallback;
            }
            else {
                ConnectionMonitor.logger.warn('Unrecognized UI update event:', update.event)
            }
        });

        if (messageCount > 0) {
            ConnectionMonitor.logger.debug(`Sending connect event for ${mapId}`);
            lastConnectCb();
        }
        else if (messageCount < 0) {
            ConnectionMonitor.logger.debug(`Sending disconnect event for ${mapId}`);
            lastDisconnectCb();
        }

        ConnectionMonitor.uiUpdateMap.set(mapId, []);
    }
}

File: receivers/android/app/src/main/java/com/futo/fcast/receiver/ConnectionMonitor.kt
package com.futo.fcast.receiver

import android.util.Log
import com.futo.fcast.receiver.models.Opcode
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.net.SocketAddress
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class ConnectionMonitor(private val _scope: CoroutineScope) {
    init {
        setInterval({
            if (_backendConnections.isNotEmpty()) {
                val keys = _backendConnections.keys.toSet()
                var removeSession = false

                for (sessionId in keys) {
                    _backendConnections[sessionId]?.let {
                        val version = it.getSessionProtocolVersion(sessionId)

                        if (version != null && version >= 2) {
                            if (_heartbeatRetries.getOrDefault(sessionId, 0) > 3) {
                                Log.w(
                                    TAG,
                                    "Could not ping device with connection id $sessionId. Disconnecting..."
                                )
                                it.disconnect(sessionId)
                                continue
                            }

                            _scope.launch(Dispatchers.IO) {
                                try {
                                    Log.d(
                                        TAG,
                                        "Pinging session $sessionId with ${_heartbeatRetries[sessionId]} retries left"
                                    )
                                    it.send(Opcode.Ping, null, sessionId)
                                    _heartbeatRetries[sessionId] =
                                        _heartbeatRetries.getOrDefault(sessionId, 0) + 1
                                } catch (e: Throwable) {
                                    Log.w(TAG, "Failed to ping session $sessionId", e)
                                }
                            }
                        } else if (version == null) {
                            Log.w(
                                TAG,
                                "Session $sessionId was not found in the list of active sessions. Removing..."
                            )
                            removeSession = true
                            _heartbeatRetries.remove(sessionId)
                        }
                    }

                    if (removeSession) {
                        _backendConnections.remove(sessionId)
                    }
                }
            }
        }, CONNECTION_PING_TIMEOUT)
    }

    companion object {
        private const val TAG = "ConnectionMonitor"

        private const val CONNECTION_PING_TIMEOUT = 2500L
        private val _heartbeatRetries = ConcurrentHashMap<UUID, Int>()
        private val _backendConnections = ConcurrentHashMap<UUID, ListenerService>()
        private const val UI_CONNECT_UPDATE_TIMEOUT = 100L
        private const val UI_DISCONNECT_UPDATE_TIMEOUT =
            2000L // Senders may reconnect, but generally need more time
        private val _uiUpdateMap =
            mutableMapOf<SocketAddress, ArrayList<Pair<String, (() -> Unit)>>>()

        fun onPingPong(sessionId: UUID) {
            Log.d(TAG, "Received response from $sessionId")
            _heartbeatRetries[sessionId] = 0
        }

        fun onConnect(
            listener: ListenerService,
            sessionId: UUID,
            address: SocketAddress,
            uiUpdateCallback: () -> Unit
        ) {
            Log.i(TAG, "Device connected: sessionId=$sessionId, address=$address")

            _backendConnections[sessionId] = listener
            _heartbeatRetries[sessionId] = 0

            // Occasionally senders seem to instantaneously disconnect and reconnect, so suppress those ui updates
            val senderUpdateQueue = _uiUpdateMap.getOrDefault(address, arrayListOf())
            senderUpdateQueue.add(Pair("connect", uiUpdateCallback))
            _uiUpdateMap[address] = senderUpdateQueue

            if (senderUpdateQueue.size == 1) {
                setTimeout({ processUiUpdateCallbacks(address) }, UI_CONNECT_UPDATE_TIMEOUT)
            }
        }

        fun onDisconnect(sessionId: UUID, address: SocketAddress, uiUpdateCallback: () -> Unit) {
            Log.i(TAG, "Device disconnected: sessionId=$sessionId, address=$address")

            _backendConnections.remove(sessionId)
            _heartbeatRetries.remove(sessionId)

            val senderUpdateQueue = _uiUpdateMap.getOrDefault(address, arrayListOf())
            senderUpdateQueue.add(Pair("disconnect", uiUpdateCallback))
            _uiUpdateMap[address] = senderUpdateQueue

            if (senderUpdateQueue.size == 1) {
                setTimeout({ processUiUpdateCallbacks(address) }, UI_DISCONNECT_UPDATE_TIMEOUT)
            }
        }

        private fun processUiUpdateCallbacks(address: SocketAddress) {
            val updateQueue = _uiUpdateMap.getOrDefault(address, arrayListOf())
            var lastConnectCb: () -> Unit = {}
            var lastDisconnectCb: () -> Unit = {}
            var messageCount = 0

            for (update in updateQueue) {
                Log.d(TAG, "Processing update event '${update.first}' for $address")
                when (update.first) {
                    "connect" -> {
                        messageCount += 1
                        lastConnectCb = update.second
                    }

                    "disconnect" -> {
                        messageCount -= 1
                        lastDisconnectCb = update.second
                    }

                    else -> {
                        Log.w(TAG, "Unrecognized UI update event: ${update.first}")
                    }
                }
            }

            if (messageCount > 0) {
                Log.d(TAG, "Sending connect event for $address")
                lastConnectCb()
            } else if (messageCount < 0) {
                Log.d(TAG, "Sending disconnect event for $address")
                lastDisconnectCb()
            }

            _uiUpdateMap[address] = arrayListOf()
        }
    }
}
